### PR TITLE
[FSDP] Decrease window size for race condition

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#76687 [FSDP] Decrease window size for race condition**
* #76665 [FSDP] Faster dict inversion
* #76664 [FSDP] Validate exec order using `compute_device`

Differential Revision: [D36074037](https://our.internmc.facebook.com/intern/diff/D36074037)

### CODE REVIEW

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36074037).

**Review by zhaojuanmao — APPROVED** [review]

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36074037).

**Comment by awgu** [comment]
@awgu has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D36074037).

File: torch/distributed/fsdp/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -2106,6 +2106,11 @@ def _pre_backward_hook(*unused: Any) -> None:
             # Run ``_pre_backward_hook`` only once per backward pass
             if self._pre_backward_hook_has_run:
                 return
+            # TODO: (Race) If two autograd threads enter this function and see
```
Comment (mrshenli) [inline torch/distributed/fsdp/fully_sharded_data_parallel.py:2109]: Curious, do we have multiple threads per device? I might miss some recent progress on autograd, I recall it was only one thread per device before?

cc @albanD

Comment (awgu) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2109]: After some local experimentation, I think we were mistaken about the race, so this PR should not be merged. One thread per device makes sense for what I am seeing empirically in my local experiments.

**Comment by awgu** [comment]
Closing this since the change is not meaningful.

File: torch/distributed/fsdp/fully_sharded_data_parallel.py (hunk 1 above)
Comment (albanD) [reply torch/distributed/fsdp/fully_sharded_data_parallel.py:2109]: Every device has a single thread associated to it indeed.
A "device" is defined as anything that is not cpu/lazy/meta.
So if you're running on cpu, you can actually have a race as two cpu thread can be working on two independent backward calls at the same time.

**Pull request closed without merging**